backfill CHANGES file

tobydickenson opened the pull request:
# Description

The CHANGES.md file has not been updated in the last few years. This PR fills in the gaps.

* Each change has a summary taken from a git commit message or Pull Request
* That summary is a single sentence. Previously each entry in CHANGES.md would have a description paragraph, but a more concise presentation makes sense because there is a large number of changes.  Most items have a PR reference for further details.
* Grammar and punctuation has been made consistent.
* I was looking out for any changes with known backwards compatibility issues so that they could have additional emphasis in this list. None were found.
* I reviewed the git log back to 2023, which is the last update recorded in CHANGES.md.
* CHANGES.md had a small number of entries from shortly after the `test` / `develop` branch split. Those entries remain in CHANGES.md as originally written.
* Users want to know what would change if they upgraded. For users upgrading from a previous 'test' branch snapshot that baseline is best identified by date, not version number.  I have therefore grouped changes into 3 month "quarterly" headings.
* I omitted changes which are "fixes to my previous fix" churn.
* I omitted changes which are not a concern for the typical CHANGES.md audience: log file content, UI wording, better error messages. These are all valuable contributions, but not a concern for someone upgrading their machine.
* I truncated the file to remove changes older than 2 years. Ancient history is available in git.

# Justification
Users need to be informed of benefits & risks when they consider upgrading their software. 

# Instructions for Use
n/a

# Implementation Details
1. How did you test the change? Be descriptive. Untested code will not be accepted. -- nothing to test
2. Did you follow the [coding style](https://github.com/openpnp/openpnp/wiki/Developers-Guide#coding-style)? -- The previous CHANGES.md content had some inconsistent styling. Feedback is welcome.
3. If you made changes in the `org.openpnp.spi` or `org.openpnp.model` packages you will need to add additional justification for these changes. Changes to these packages require extensive review and testing. -- no changes
4. Be sure to run `mvn test` before submitting the Pull Request. If the tests do not pass the Pull Request will not be accepted. -- there are no code changes

janm012012 commented:
Hi Toby!

Nice work! That's very valuable for users considering to updated or no. Thank You!

Jan

ian-arkver commented:
I was a little surprised to see all the old history gone, but you're right, it's still in git if needed. Possibly an OLD_CHANGES.md or something might have kept it more visible.

The diff is hard to read. Easier to review the finished file here:
https://github.com/tobydickenson/openpnp/blob/update-changes/CHANGES.md

Quite the nice cleanup. Looks good to me.

atanisoft commented:
I agree, this looks a lot better. One small comment though, the `#1234` numbers at the ends of lines could be considered either an issue number or PR number. It would be good to clarify as `PR #1234` when it is a PR vs when it is an issue. Alternatively add a note at the top that these are PRs and not issues, if that is what they all are.

tobydickenson commented:
1. I changed the `#1234` references to links. These will get rendered as a clickable  `#1234` when merged into openpnp git.
2. I added a link to todays CHANGES.md under an "Old Changes" heading.

ian-arkver commented:
Looks even better to me. Thanks!

janm012012 commented:
Hi Toby!

To boost OpenPnP development, I'd like to get this PR merged before deploying the a stable version. I think it's good as it is, probably a reference to the G-Code fix merged yesterday by PR #1728 would be nice.

Any amendments or concerns?

Jan

tobydickenson commented:
> To boost OpenPnP development, I'd like to get this PR merged before deploying the a stable version. I think it's good as it is
> Any amendments or concerns?

It looks ready to me :-)

> probably a reference to the G-Code fix merged yesterday by PR #1728 would be nice.

I havent yet got back to my desk to check the details of that fix, but I'm guessing it was a fix for a bug I introduced. If so,
a. Thank you for the fix!
b. I dont think a changelog needs to separately mention the fixes for bugs in new features which are described elsewhere in that changelog.

Toby

jbussmann commented:
Awesome write up, I'm excited to test all the new features, as soon as I get a new P&P job. :)

One small note, several hours too late: Optimize placements of multi nozzle machines #1574 is mentioned twice.

janm012012 commented:
Hi @tobydickenson!

I just tried to fix the double #1574 @jbussmann reported and noticed, that the file looks quite different in the Eclipse preview. Digging into the source, I noticed, that you added the URLs to each entry as string without syntax. In the Eclipse previewer they appear as just a sequence of characters, while on github (where I checked the file before merging) they are interpreted as links and show up abbreviated as `#<PR number>`. (IIRC the same magic is applied when viewing the file in the installer). I'm not sure this behavior will stay like that in the future and if we shall relay on that and would therefore kindly ask you to change them to Makedown syntax (`[<visible name>](<url>)`). Then they look the same in the preview as on github and no interpretation will/shall take place. Thank You! (I'd love to merge such a PR before creating the next stable version.)

Jan

tobydickenson commented:
Duplicate item removed

PR link markdown added using some sed magic:

`
sed -i -e "s,\(https://github.com/openpnp/openpnp/pull/\([0-9]*\)\),[PR \2](\1),g" CHANGES.md
`

This is on my original update-changes branch. Do I need to open another PR since this one is merged?

janm012012 commented:
Hi Toby!

Many thanks! I can't see you commit and I don't know if we can amend something. You may try and we'll see. If you can't push, you'll need to open a new PR.

Thank You for changing this!

Jan

tobydickenson commented:
> you'll need to open a new PR.

Please see #1734
